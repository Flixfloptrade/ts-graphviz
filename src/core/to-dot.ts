import { DotObjectModel } from '../common/index.js';
import { stringify, fromModel, ConvertFromModelOptions, SerializeOptions } from '../ast/index.js';

/**
 * This interface provides options for converting a model to DOT.
 * @group Convert Model to DOT
 * @alpha
 */
export interface ToDotOptions {
  /**
   * Options for converting the model to DOT.
   */
  convert?: ConvertFromModelOptions;
  /**
   * Options for printing DOT.
   */
  print?: SerializeOptions;
}

/**
 * Convert Model to DOT string.
 *
 * @group Convert Model to DOT
 *
 * @param model Dot Object Model, like Digraph, Graph, Subgraph, Node, and Edge.
 * @param options Optional options for the conversion.
 * @returns DOT string
 */
export function toDot(model: DotObjectModel, options?: ToDotOptions): string {
  const ast = fromModel(model, options?.convert);
  return stringify(ast, options?.print);
}
