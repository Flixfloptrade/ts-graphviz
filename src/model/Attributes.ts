import { DotBase } from '../common';
import { indent } from '../utils/dot-rendering';
import { Literal } from './Literal';

/**
 * @category Attributes
 */
export class Attributes extends DotBase {
  protected attrs: Map<string, Literal> = new Map();
  get size(): number {
    return this.attrs.size;
  }
  public get(key: string): Literal | undefined {
    return this.attrs.get(key);
  }
  public set(key: string, value: any): void {
    if (value instanceof Literal) {
      this.attrs.set(key, value);
    } else {
      this.attrs.set(key, new Literal(value));
    }
  }
  public toDot(): string {
    if (this.size === 0) {
      return '';
    }
    return [
      '[',
      ...Array.from(this.attrs.entries()).map(([key, value]) => indent(`${key} = ${value.toDot()},`)),
      ']',
    ].join('\n');
  }
}
