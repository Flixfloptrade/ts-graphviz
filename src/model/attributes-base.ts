import { DotObject } from './abstract';
import { AttributesObject, AttributesValue, IAttributesBase, IAttributes } from '../types';

/**
 * @hidden
 */
export abstract class AttributesBase<T extends string> extends DotObject implements IAttributesBase<T> {
  /** @hidden */
  protected attrs: Map<T, AttributesValue> = new Map();

  public entries(): IterableIterator<[T, AttributesValue]> {
    return this.attrs.entries();
  }

  /** The size of the attribute. */
  get size(): number {
    return this.attrs.size;
  }
  /** The size of the attribute. */
  public get(key: T): AttributesValue | undefined {
    return this.attrs.get(key);
  }
  /** Set a value to the attribute. */
  public set(key: T, value: AttributesValue): void {
    this.attrs.set(key, value);
  }

  public delete(key: T): void {
    this.attrs.delete(key);
  }

  public apply(attributes: AttributesObject<T>): void {
    for (const [key, value] of Object.entries(attributes)) {
      this.set(key as T, value as string | boolean | number);
    }
  }

  public clear(): void {
    this.attrs.clear();
  }
}

/**
 * A set of attribute values for any object.
 *
 * @category Attributes
 */
export class Attributes<T extends string = string> extends AttributesBase<T> implements IAttributes<T> {
  /** Comments to include when outputting with toDot. */
  public comment?: string;
}
