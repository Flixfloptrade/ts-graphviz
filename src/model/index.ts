export * from './Attributes';
export * from './Cluster';
export * from './RootCluster';
export * from './Graph';
export * from './Digraph';
export * from './Edge';
export * from './Node';
export * from '../render/Context';
export * from '../render/to-dot';
