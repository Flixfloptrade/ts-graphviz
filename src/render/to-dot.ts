/* eslint-disable @typescript-eslint/no-use-before-define */
import { Renderer } from './renderer';
import { IContext } from '../types';

export function toDot(object: unknown, context?: IContext): string | undefined {
  const renderer = new Renderer(context);
  return renderer.render(object);
}
