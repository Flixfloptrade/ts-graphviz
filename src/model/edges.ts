import { DotObject } from './abstract';
import { attribute } from '../attribute';
import { EdgeTarget } from '../types';
import { Attributes } from './attributes-base';

/**
 * @category Primary
 */
export class Edge extends DotObject {
  /** Comments to include when outputting with toDot. */
  public comment?: string;
  // TODO
  public readonly attributes = new Attributes<attribute.Edge>();

  constructor(public readonly targets: EdgeTarget[]) {
    super();
  }
}
