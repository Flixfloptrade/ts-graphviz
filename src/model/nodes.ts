import { NodeRefLike, NodeRef, INode, IPort, IForwardRefNode, NodeAttributes, IAttributes } from '../types';
import { DotObject } from './abstract';
import { attribute } from '../attribute';
import { Attributes } from './attributes-base';
import { NodeRefGroupLike } from '../types';

/**
 * Node object.
 * @category Primary
 */
export class Node extends DotObject implements INode {
  /** Comments to include when outputting with toDot. */
  public comment?: string;
  public readonly attributes: IAttributes<attribute.Node>;
  constructor(public readonly id: string, attributes?: NodeAttributes) {
    super();
    this.attributes = new Attributes<attribute.Node>(attributes);
  }

  /** Returns ForwardRefNode with port and compass specified. */
  public port(port: string | Partial<IPort>): IForwardRefNode {
    if (typeof port === 'string') {
      return { id: this.id, port };
    }
    return { id: this.id, ...port };
  }
}

/** @hidden */
export function isForwardRefNode(object: unknown): object is IForwardRefNode {
  return typeof object == 'object' && object !== null && typeof (object as IForwardRefNode).id === 'string';
}

/** @hidden */
export function isEdgeTarget(node: unknown): node is NodeRef {
  return node instanceof Node || isForwardRefNode(node);
}

/** @hidden */
export function isNodeRefLike(node: unknown): node is NodeRefLike {
  return typeof node === 'string' || isEdgeTarget(node);
}

/** @hidden */
export function isNodeRefGroupLike(target: NodeRefLike | NodeRefGroupLike): target is NodeRefGroupLike {
  return Array.isArray(target) && target.every(isNodeRefLike);
}
