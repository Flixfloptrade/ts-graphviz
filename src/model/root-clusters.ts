import { attribute } from '../attribute';
import { IRootCluster } from '../types';
import { Cluster } from './clusters';
import { Attributes } from './attributes-base';

/**
 * Base class for RootCluster.
 */
export abstract class RootCluster extends Cluster<attribute.RootCluster> implements IRootCluster {
  /**
   * Strict mode.
   *
   * @description
   * A graph may also be described as strict.
   * This forbids the creation of multi-edges, i.e., there can be at most one edge with a given tail node and head node in the directed case.
   * For undirected graphs, there can be at most one edge connected to the same two nodes.
   * Subsequent edge statements using the same two nodes will identify the edge with the previously defined one and apply any attributes given in the edge statement.
   */
  public strict = false;

  public attributes = {
    graph: new Attributes<attribute.Subgraph | attribute.ClusterSubgraph>(),
    edge: new Attributes<attribute.Edge>(),
    node: new Attributes<attribute.Node>(),
  };
  constructor(public readonly id?: string) {
    super();
  }
}

/**
 * Graph object.
 *
 * @description
 * An object representing an omnidirectional graph.
 * @category Primary
 */
export class Graph extends RootCluster {}

/**
 * Digraph object.
 *
 * @description
 * The object representing a directional graph.
 * @category Primary
 */
export class Digraph extends RootCluster {}
