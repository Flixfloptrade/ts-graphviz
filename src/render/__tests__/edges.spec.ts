import { Edge } from '../../model/edges';
import { Node } from '../../model/nodes';
import { Digraph, Graph } from '../../model/root-clusters';
import { toDot } from '../to-dot';

describe('Edge rendering', () => {
  let edge: Edge;

  const [node1, node2, node3] = Array(3)
    .fill(true)
    .map((_, i) => new Node(`node${i + 1}`));

  beforeEach(() => {
    edge = new Edge([node1, node2]);
  });

  describe.each([
    [
      'Digraph context',
      {
        root: new Digraph(),
      },
    ],
    [
      'Graph context',
      {
        root: new Graph(),
      },
    ],
  ])('%s', (_, context) => {
    it('simple edge', () => {
      expect(toDot(edge, context)).toMatchSnapshot();
    });

    describe('edge with comment', () => {
      test('single line comment', () => {
        edge.comment = 'this is comment.';
        expect(toDot(edge, context)).toMatchSnapshot();
      });

      test('multi line comment', () => {
        edge.comment = 'this is comment.\nsecond line.';
        expect(toDot(edge, context)).toMatchSnapshot();
      });
    });

    it('has some attributes', () => {
      edge.attributes.set('label', 'this is test');
      edge.attributes.set('color', 'red');
      expect(toDot(edge, context)).toMatchSnapshot();
    });

    it('3 nodes', () => {
      edge = new Edge([node1, node2, node3]);
      expect(toDot(edge, context)).toMatchSnapshot();
    });

    it('3 nodes, but many args', () => {
      edge = new Edge([node1, node2, node3, node1, node2, node3, node1, node2, node3]);
      expect(toDot(edge, context)).toMatchSnapshot();
    });
  });
});
