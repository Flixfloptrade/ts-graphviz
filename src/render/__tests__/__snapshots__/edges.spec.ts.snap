// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edge rendering Digraph context 3 nodes 1`] = `"\\"node1\\" -> \\"node2\\" -> \\"node3\\";"`;

exports[`Edge rendering Digraph context 3 nodes, but many args 1`] = `"\\"node1\\" -> \\"node2\\" -> \\"node3\\" -> \\"node1\\" -> \\"node2\\" -> \\"node3\\" -> \\"node1\\" -> \\"node2\\" -> \\"node3\\";"`;

exports[`Edge rendering Digraph context edge with comment multi line comment 1`] = `
"// this is comment.
// second line.
\\"node1\\" -> \\"node2\\";"
`;

exports[`Edge rendering Digraph context edge with comment single line comment 1`] = `
"// this is comment.
\\"node1\\" -> \\"node2\\";"
`;

exports[`Edge rendering Digraph context has some attributes 1`] = `
"\\"node1\\" -> \\"node2\\" [
  label = \\"this is test\\",
  color = \\"red\\",
];"
`;

exports[`Edge rendering Digraph context simple edge 1`] = `"\\"node1\\" -> \\"node2\\";"`;

exports[`Edge rendering Graph context 3 nodes 1`] = `"\\"node1\\" -- \\"node2\\" -- \\"node3\\";"`;

exports[`Edge rendering Graph context 3 nodes, but many args 1`] = `"\\"node1\\" -- \\"node2\\" -- \\"node3\\" -- \\"node1\\" -- \\"node2\\" -- \\"node3\\" -- \\"node1\\" -- \\"node2\\" -- \\"node3\\";"`;

exports[`Edge rendering Graph context edge with comment multi line comment 1`] = `
"// this is comment.
// second line.
\\"node1\\" -- \\"node2\\";"
`;

exports[`Edge rendering Graph context edge with comment single line comment 1`] = `
"// this is comment.
\\"node1\\" -- \\"node2\\";"
`;

exports[`Edge rendering Graph context has some attributes 1`] = `
"\\"node1\\" -- \\"node2\\" [
  label = \\"this is test\\",
  color = \\"red\\",
];"
`;

exports[`Edge rendering Graph context simple edge 1`] = `"\\"node1\\" -- \\"node2\\";"`;
